coderd/provisionerdserver/provisionerdserver_test.go

```diff
@@ -1804,6 +1804,155 @@ func TestNotifications(t *testing.T) {
 			})
 		}
 	})
+
+	t.Run("TemplateBuildFailed", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			ctx      = context.Background()
+			notifEnq = &testutil.FakeNotificationsEnqueuer{}
+			// To avoid spamming the output, ignore log errors. This test is
+			// designed to check a build failure, which is expected to log errors.
+			ignoreLogErrors = true
+			srv, db, ps, pd = setup(t, ignoreLogErrors, &overrides{
+				notificationEnqueuer: notifEnq,
+			})
+			// Create users with different roles to ensure that notifications are sent
+			// to the appropriate users
+			ownerA   = dbgen.User(t, db, database.User{RBACRoles: []string{codersdk.RoleOwner}})
+			ownerB   = dbgen.User(t, db, database.User{RBACRoles: []string{codersdk.RoleOwner}})
+			tplAdmin = dbgen.User(t, db, database.User{RBACRoles: []string{codersdk.RoleTemplateAdmin}})
+			_        = dbgen.User(t, db, database.User{RBACRoles: []string{codersdk.RoleUserAdmin}})
+			member   = dbgen.User(t, db, database.User{RBACRoles: []string{codersdk.RoleMember}})
+		)
+
+		tc := []struct {
+			name           string
+			tplOwner       database.User
+			buildInitiator database.User
+			reason         database.BuildReason
+			transition     database.WorkspaceTransition
+			receivers      []uuid.UUID
+		}{
+			{
+				name:           "ManualBuild",
+				tplOwner:       ownerA,
+				buildInitiator: ownerB,
+				reason:         database.BuildReasonInitiator,
+				transition:     database.WorkspaceTransitionStart,
+				// Ensure that during manual builds, the initiator does not receive a
+				// notification. In this scenario, ownerB should not receive a
+				// notification.
+				receivers: []uuid.UUID{ownerA.ID, tplAdmin.ID},
+			},
+			{
+				name:           "AutoBuild",
+				tplOwner:       ownerA,
+				buildInitiator: ownerA,
+				reason:         database.BuildReasonAutostart,
+				transition:     database.WorkspaceTransitionStart,
+				// Ensure that during automated builds, all template admins and owners
+				// receive notifications.
+				receivers: []uuid.UUID{ownerA.ID, ownerB.ID, tplAdmin.ID},
+			},
+		}
+
+		for _, c := range tc {
+			t.Run(c.name, func(t *testing.T) {
+				// Given: a template and a workspace build
+				isManualBuild := c.reason == database.BuildReasonInitiator
```

Comment:
Try and group your code into logical blocks, and use whitespace for visual assistance. Think of your code as having the same visual hierarchy that you'd design in your frontends.

For example, `isManualBuild` is only needed lower down: group them logically.

You've got a whole lot of code below which is doing several things; add a bit of spaces and/or comments to give the reader's "eye" some space to "breathe", and group related concepts/lines together to infer relationships.